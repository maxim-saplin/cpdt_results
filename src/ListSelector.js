import React, { Component } from 'react';
import l18n from './translations';

class ListSelector extends Component {
  constructor(props) {
    super(props);

    this.toggle = this.toggle.bind(this);
    this.selectAll = this.selectAll.bind(this);
    this.onKeyPress = this.onKeyPress.bind(this);
  }

  toggle(key) {
    if (this.props.selectedKey.includes(key)) { // remove key
      if (this.props.selectedKey.length > 1) {
        return this.props.selectedKey.filter(k => k !== key);
      }
      else {
        return this.props.selectedKey;
      }
    }

    this.props.selectedKey.push(key);

    return this.props.selectedKey;
  }

  selectAll() {
    return this.props.items.map(p => p.key);
  }

  onKeyPress(e) {
    if (document.activeElement instanceof HTMLBodyElement){
      if (this.shortcuts.has(e.key))
      {
        let val = this.shortcuts.get(e.key);

        //TODO - add suport for repeting shortcuts in multiselect list
        if (Array.isArray(val)){ // single select, repeat shortcuts
          let key = val[0].key;
          
          for (let i = 0; i < val.length; i++){
            if (val[i].key === this.props.selectedKey){
              if (i+1 < val.length) key = val[i+1].key;
            }
          }

          this.props.itemClick(key);
        }
        else if (Array.isArray(this.props.selectedKey)){ // multi select select, non-repeat shortcuts        
          this.props.itemClick(this.toggle(val.key));
        } else if (val.key !== this.props.selectedKey) { // single select, non-repeat shortcuts
          this.props.itemClick(val.key);
        } 
      }
    }
  }

  componentWillUnmount(){
    if (this.keyPressSubscribed)
      document.removeEventListener('keypress', this.onKeyPress);
  }

  initShortcuts() {
    this.prevItems = this.props.items;
    this.shortcuts = new Map();
    for (var i in this.prevItems) {
      if (this.prevItems[i].shortcut) {
        if (!this.shortcuts.has(this.prevItems[i].shortcut))
          this.shortcuts.set(this.prevItems[i].shortcut, this.prevItems[i]);
        else {
          let sc = this.shortcuts.get(this.prevItems[i].shortcut);

          if (Array.isArray(sc)) sc.push(this.prevItems[i]);
          else {
            let arr = [sc, this.prevItems[i]];
            this.shortcuts.set(this.prevItems[i].shortcut, arr);
          }
        }
      }
    }
    if (this.shortcuts.size > 0) {
      document.addEventListener('keypress', this.onKeyPress);
      this.keyPressSubscribed = true;
    }
    else if (this.keyPressSubscribed) {
      document.removeEventListener('keypress', this.onKeyPress);
      this.keyPressSubscribed = false;
    }
  }

  updateShortcuts() {
    if (!this.prevItems) {
      this.initShortcuts();
    }
    else if (this.prevItems !== this.props.items) {
      this.initShortcuts();
    }
  }

  render() {
    this.updateShortcuts();

    if (this.props.items)
      return (
        <ul className={this.props.className}>
          {
              this.props.items.map( // multiselect
                i =>
                  <li
                    key={i.key}
                    className={this.props.selectedKey.includes(i.key) ? "selected" : null}
                    onClick={
                      Array.isArray(this.props.selectedKey) ? 
                      () => this.props.itemClick(this.toggle(i.key)) : // multiselect
                      () => this.props.itemClick(i.key) //single select
                    }
                  >
                    {i.name}
                  </li>
              )
          }

          {
            this.props.selectAll && <li onClick={() => this.props.itemClick(this.selectAll())}>{l18n.all}</li>
          }
        </ul>
      )

      return null;
  }
}



export default ListSelector;