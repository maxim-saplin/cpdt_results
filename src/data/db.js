import dictionaries from './dictionaries';
import results from './results';

const db_version_url = "https://raw.githubusercontent.com/maxim-saplin/cpdt_results/master/Results_version.json";
const db_url = "https://raw.githubusercontent.com/maxim-saplin/cpdt_results/master/Results.json";
const db_version_sotrage_key = "cpdt_db_version";
const db_sotrage_key = "cpdt_db";

const db = {
    dictionaries : dictionaries,
    results : results,
    
    init: function (){
        let fetchVersion = () => {
            return new Promise (resolve => {
                fetch(db_version_url)
                .then(response => {
                    return response.json();
                })
                .then(result => {
                    storage.setItem(db_version_sotrage_key, result.v);
                    this.version = result.v;
                    resolve(result.v);
                });
            });
        };

        let fetchDb = () => {
            return new Promise (resolve => {
                fetch(db_url)
                .then(response => response.json())
                .then(result => {
                    storage.setItem(db_sotrage_key, JSON.stringify(result));
                    dictionaries.data = result;
                    results.data = result;
                    resolve(result);
                })
            });
        };
        
        let promise = null;

        let storage = window.localStorage;
        let data = storage.getItem(db_sotrage_key);
        if (data) {
          try { data = JSON.parse(data);} catch{};
          if (!data.results) data = null; // shallow check for corrupt data
        }
        let localDbVersion = storage.getItem(db_version_sotrage_key);

        if (data){
            promise = new Promise( resolve => {
                fetchVersion()
                .then(result => {
                    if (result !== localDbVersion){
                        fetchDb().then(result => resolve(result));                    
                    }
                    // local data is up to date
                    else{ 
                        dictionaries.data = data;
                        results.data = data;
                        resolve(data)
                    };
                });
            });
        }
        else{
            fetchVersion();
            promise = fetchDb();
        }

        return promise;
    },

    initFake: function() {
      dictionaries.data = {"results":[{"id":"1","device":"Xiaomi Mi8 SE","platfrom":"Android","deviceDetail":"Snapdragon 710, 64GB","deviceComment":"","deviceYear":"2018","dateTested":"23.10.2018","seqWrite":"110","seqRead":"299","randWrite":"18,9","randRead":"9,5","memCopy":"4,65"},{"id":"2","device":"Samsung Galaxy S9","platfrom":"Android","deviceDetail":"Exynos 9810, 64Gb","deviceComment":"","deviceYear":"2018","dateTested":"04.10.2018","seqWrite":"118","seqRead":"654","randWrite":"1,72","randRead":"14,43","memCopy":"12,7"},{"id":"3","device":"Essential Phone PH-1","platfrom":"Android","deviceDetail":"Snapdragon 835, 128GB","deviceComment":"","deviceYear":"2017","dateTested":"30.09.2018","seqWrite":"125","seqRead":"440","randWrite":"4,3","randRead":"17,4","memCopy":"6,6"},{"id":"4","device":"15\" MacBook Pro","platfrom":"macOS","deviceDetail":"macOS, APFS","deviceComment":"","deviceYear":"2018","dateTested":"03.10.2018","seqWrite":"1331,2","seqRead":"2088,96","randWrite":"75","randRead":"38","memCopy":"10,3"},{"id":"5","device":"15\" MacBook Pro","platfrom":"macOS","deviceDetail":"macOS, NTFS partition","deviceComment":"","deviceYear":"2018","dateTested":"03.10.2018","seqWrite":"730","seqRead":"694","randWrite":"27","randRead":"8","memCopy":"10,3"},{"id":"6","device":"15\" MacBook Pro","platfrom":"Windows","deviceDetail":"Windows 10 Bootcamp, NTFS","deviceComment":"","deviceYear":"2018","dateTested":"03.10.2018","seqWrite":"343","seqRead":"1617,92","randWrite":"0,5625","randRead":"11,01","memCopy":"8,5"},{"id":"7","device":"15\" MacBook Pro","platfrom":"Windows","deviceDetail":"Windows 10, VMWare Fusion, NTFS","deviceComment":"","deviceYear":"2018","dateTested":"03.10.2018","seqWrite":"850","seqRead":"2457,6","randWrite":"12","randRead":"29","memCopy":"5,8"}],"platforms":["Android", "macOS", "Windows"],"tests":["seqWrite", "seqRead", "randWrite", "randRead", "memCopy"]};
      results.data = dictionaries.data;
    }
};

export default db;