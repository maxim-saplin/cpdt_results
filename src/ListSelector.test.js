import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
// import Adapter from 'enzyme-adapter-react-16';
// import Enzyme, { shallow } from 'enzyme';
import ListSelector from './ListSelector';
import db from './data';

//Enzyme.configure({ adapter: new Adapter() });

beforeAll(() => {
  db.initFake();
});

describe('ListSelector', () => {

  it('renders empty without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<ListSelector />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders with items without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<ListSelector  items={db.dictionaries.getTests()} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('shortcuts are created', () => {
    const component = renderer.create(
      <ListSelector items={db.dictionaries.getTests()}  />
    ).getInstance();

    expect(component.shortcuts.size).toBe(3);
  });

  it('for repeated shortcuts there are arrays values in the Map', () => {
    const component = renderer.create(
      <ListSelector items={db.dictionaries.getTests()}  />
    ).getInstance();

    //[S]equential write[S]equential read[R]andom write (4KB)[R]andom read (4KB)[M]emory copy
    expect(component.shortcuts.get('s').length).toBe(2);
    expect(component.shortcuts.get('r').length).toBe(2);
  });

  it('key press is handled without errors', () => {
    let key = "";
    
    const component = renderer.create(
      <ListSelector items={db.dictionaries.getTests()} itemClick={(k) => key = k} />
    ).getInstance();

    component.keyPress({key: 'm'});
    //[S]equential write[S]equential read[R]andom write (4KB)[R]andom read (4KB)[M]emory copy
    //[M] - m - memCopy
    expect(key).toBe('memCopy');
  });

  test('snapshots', () => {
    const component = renderer.create(
      <ListSelector items={db.dictionaries.getTests()}  />
    );

    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

});