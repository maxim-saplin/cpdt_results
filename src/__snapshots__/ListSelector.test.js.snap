// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListSelector snapshots 1`] = `
<ul>
  <li
    className={null}
    onClick={[Function]}
  >
    [S]equential write
  </li>
  <li
    className={null}
    onClick={[Function]}
  >
    [S]equential read
  </li>
  <li
    className={null}
    onClick={[Function]}
  >
    [R]andom write (4KB)
  </li>
  <li
    className={null}
    onClick={[Function]}
  >
    [R]andom read (4KB)
  </li>
  <li
    className={null}
    onClick={[Function]}
  >
    [M]emory copy
  </li>
</ul>
`;
